src/lib.rs

```diff
@@ -595,12 +642,87 @@ impl Cancelled {
 
 impl std::fmt::Display for Cancelled {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str("cancelled")
+        let why = match self {
+            Cancelled::PendingWrite => "pending write",
+            Cancelled::PropagatedPanic => "propagated panic",
+            Cancelled::UnexpectedCycle(_) => "unexpected cycle",
+        };
+        f.write_str("cancelled because of ")?;
+        f.write_str(why)
     }
 }
 
 impl std::error::Error for Cancelled {}
 
+/// Information about an "unexpected" cycle, meaning one where some of the
+/// participants lacked cycle recovery annotations.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Cycle {
+    participants: plumbing::CycleParticipants,
+}
+
+impl Cycle {
+    pub(crate) fn new(participants: plumbing::CycleParticipants) -> Self {
+        Self { participants }
+    }
+
+    /// Iterate over the [`DatabaseKeyIndex`] for each query participating
+    /// in the cycle. The start point of this iteration within the cycle
+    /// is arbitrary but deterministic, but the ordering is otherwise determined
+    /// by the execution.
+    pub fn participant_keys(&self) -> impl Iterator<Item = DatabaseKeyIndex> + '_ {
+        let min = self.participants.iter().min().unwrap();
+        let index = self.participants.iter().position(|p| p == min).unwrap();
+        self.participants[index..]
+            .iter()
+            .chain(self.participants[..index].iter())
+            .copied()
+    }
+
+    /// Returns a vector with the debug information for
+    /// all the participants in the cycle.
+    pub fn all_participants<DB: ?Sized + Database>(&self, db: &DB) -> Vec<String> {
+        self.participant_keys()
+            .map(|d| format!("{:?}", d.debug(db)))
```

Comment:
I think we have an invariant here that db's current revision matches the revision when the Cycle was created. I wonder if it makes sense to add a revision field to CycleError and a debug assert here? Or, rathre, I am fairly certain it makes sense, but I don't know whether it is worth it -- it's quitly unlikely that you'd use CycleErrors arcros revisions accidentally I would think?

Reply:
Hmm, Database key ids don't go stale-- at least, not presently. That said, if we ever did try to garbage collect keys then they *could*... really they are basically exactly like interned values (they are, indeed, interned values).

I don't think we want to assert that the cycle is only being used in the most recent revision. I feel like people may embed cycles in the recovery value and it could be possible for those to be judged as "nothing changed".